### ISSUE TITLE
Websocket (WS to WSS) reverse proxy (and/or Docker support?)

### ISSUE THREAD

**Issue by peanutlasko** [Marekkon5/onetagger#111]
Hi @Marekkon5 , I know we've chatted once before about some Onetagger questions. First off let me state that I know you guys are a small 2-man team, and expecting extensive fixes/testing/troubleshooting for **FREE** software is not something that's feasible. 

Before I dig into my issue, let me explain what I'm trying to accomplish. Right now I have a homelab/server in which I have several docker containers that run, including one that hosts my reverse proxy. All these services provide lots of different things, but one of them is for all of my audio/DJ management. 

**OneTagger is a godsend to my DJ workflow.** Right now I use it locally on my Mac Pro (server) to tag all my music, and then sync it with my Macbook Pro, which in turn I import into Rekordbox/Traktor.

Ideally, what I am trying to accomplish is the need to not be physically sitting at my Home Server. I'd like to be able to log into OneTagger via the web using the --expose --server and tag all my tracks on any PC, Anywhere. My Home server already has an automated way to use Spotify, download tracks, sort them into folders, etc. So the final piece of my puzzle is getting Onetagger to work via the web. I can do this from any computer with a web browser, securely. 

With all that out of the way, let me get a little deeper. I have a reverse proxy as I mentioned that exposes all my services. It redirects all HTTP traffic to HTTPS. I've got about 20ish applications running that work perfectly fine with this. However, with Onetagger an issue arises because of the insecure WS socket connection. I've dug into this and found several ways to mitigate the issue ([here](https://community.traefik.io/t/websocket-does-not-upgrade-ws-to-wss/12017))

The UI loads, but after a moment I'm getting a websocket connection ([screenshot here](https://imgur.com/NSMgVwc))

Now, let me be the first to say - i understand this is a pretty fringe use-case. My other alternative I've tried was to create a docker container (which I was about 90% successful), however I ran into a problem when clicking the "Browse" buttons because I guess building it from source there's no way to display via the web a dialogue box? The error I got on the docker container was:

```
missing software! (we will try basic console input)


 ___________

/           \ 

| tiny file |

|  dialogs  |

\_____  ____/

      \|

tiny file dialogs on UNIX needs:

   applescript or kdialog or yad or Xdialog

or zenity (or matedialog or shellementary or qarma)

or python (2 or 3) + tkinter + python-dbus (optional)

or dialog (opens console if needed) ** Disabled by default **/

or xterm + bash (opens console for basic input)

or existing console for basic input
```

I tried installing python3 and tkinter and python-dbus, with no such luck.

In any case, I'm hoping to either A) get a HTTPS/WSS socket that works which I can use for my reverse proxy, or B) some effort into the Docker container (which I have some experience with and would be happy to perform testing and updates).

Now all that being said, what I'm asking for is a lot. I don't expect ANY of this for free as a matter of fact. I'm going to submit a donation now, and additionally if you guys decide to tackle this I'm happy to continue to compensate you for your efforts.

**Comment by Marekkon5** [comment]
Hello, thanks for trying to use this project in such a way, however there are some issues:

1. The socket not connecting error - I will have to probably try to recreate your setup and try it myself. I am pretty certain the issue is somewhere with the socket not being properly forwarded / decrypted to HTTP.
2.  As for the docker - well I recommend just using some latest Ubuntu base image and downloading the binary (and dependencies). I don't think there needs to be any advanced/complicated setup for this.
3. As for the browse button - 1T was designed to be used on desktop, so the `tinyfiledialogs` library opens the native file browser (which on server isn't available, so it shows you that long error). This requires me to make my own file browser within 1T, which will take me some time and effort.
4. Player in QT won't work for you at all - it is server side, the music would play from the server, not the browser, and there isn't much I can do, because of the codecs support.

**Comment by peanutlasko** [comment]
Thanks for the feedback. One other log I forgot to attach was the console output from OneTagger when attempting to hit the page:

```
2022-05-06 15:20:16 [ERROR] onetagger_shared: PANIC: panicked at 'called `Result::unwrap()` on an `Err` value: HandshakeError::Failure(Protocol(HttparseError(Token)))', crates/onetagger/src/socket.rs:151:57
2022-05-06 15:20:16 [ERROR] onetagger_shared: LOCATION: File: crates/onetagger/src/socket.rs, Line: 151
```

My docker container works perfectly, its just the file browser issue that's the problem (and the websocket).

I see what your saying now though - when I access via web on my macbook (not running 1T) and hit the play button, the audio comes out of my Mac Pro (server running 1T). That's too bad. 

So even if the socket connection were fixed, and/or a docker container, I still couldn't listen to my tracks remotely and tag them. Oh well, maybe someday for a v2.0 release or something. I'll just have to live :)

**Comment by Marekkon5** [comment]
Hello, in the latest commit I've added: 
1. Client side file browser - it's been planned for some time, so I just added it.
2. Client side music player. However because the browser doesn't support all the codecs 1T does, 1T internally converts the audio file into WAV, so the browser can stream it. This causes some small delays when playing and seeking, however it enables you to play all the audio files remotely.

As for the secure sockets - I think issue is on your end with your reverse proxy setup. Do you mind sharing what programs/configs do you use?

Also to test the changes you can get the binary from the Actions tab. Let me know if it works for you. Thanks

**Comment by peanutlasko** [comment]
Hey thanks for this update!

I'm using Traefik as my reverse proxy. I can share some of my configuration with you, hopefully it makes sense. 

First I defined my entry point in Traefik.yml

```
entrypoints:
  http:
    address: ":80"
  https:
    address: ":443"
  ws:
    address: ":36912"
```

Next, I have my middlewares configured for all my services which controls headers:

```
    middlewares-secure-headers:
      headers:
        accessControlAllowMethods:
          - GET
          - OPTIONS
          - PUT
        accessControlMaxAge: 100
        hostsProxyHeaders:
          - "X-Forwarded-Host"
          - "X-Forwarded-Proto"
          #- "X-Forwarded-For"
        stsSeconds: 63072000
        #sslProxyHeaders:
          #X-Forwarded-Proto: "https"
        stsIncludeSubdomains: true
        stsPreload: true
        forceSTSHeader: true
        # frameDeny: true #overwritten by customFrameOptionsValue
        #customFrameOptionsValue: "allow-from https:example.com" #CSP takes care of this but may be needed for organizr.
        contentTypeNosniff: true
        browserXssFilter: true
        # sslForceHost: true # add sslHost to all of the services
        # sslHost: "example.com"
        referrerPolicy: "same-origin"
        # Setting contentSecurityPolicy is more secure but it can break things. Proper auth will reduce the risk.
        # the below line also breaks some apps due to 'none' - sonarr, radarr, etc.
        # contentSecurityPolicy: "frame-ancestors '*.example.com:*';object-src 'none';script-src 'none';"
        #permissionsPolicy: "camera=()" #"; geolocation=(); microphone=(); payment=(); usb=(); vr=()"
        FeaturePolicy: "camera 'none'; geolocation 'none'; microphone 'none'; payment 'none'; usb 'none'; vr 'none';"
        customResponseHeaders:
          Content-Security-Policy: "upgrade-insecure-requests"
          #Content-Security-Policy-Report-Only: "upgrade-insecure-requests"
          X-Robots-Tag: "none,noarchive,nosnippet,notranslate,noimageindex,"
          server: ""
        customRequestHeaders:
          X-Forwarded-Proto: "https,wss,ws"
```

To note, I had to add the "wss,ws" for the CustomRequestHeaders to get the GUI to load. 

Last, I have a custom rule for Onetagger (right now I'm not using a docker container, I'm just testing running it locally on my Mac)

```
http:
  routers:
    onetagger-rtr:
      rule: "Host(`onetagger.xxxxxx.net`)" 
      entryPoints:
        - https
      middlewares:
        - chain-no-auth
      service: onetagger-svc
      tls:
        certResolver: dns-cloudflare

    onetagger-rtr-ws:
      rule: "Host(`onetagger.xxxxxxx.net`)"
      entryPoints:
        - ws
      middlewares:
        - chain-no-auth
      service: onetagger-svc-ws
      tls:
        certResolver: dns-cloudflare


  services:
    onetagger-svc:
      loadBalancer:
        servers:
          - url: "http://192.168.1.101:36913"  # or whatever your external host's IP:port is 
    onetagger-svc-ws:
      loadBalancer:
        sticky:
          cookie: {}
        servers:
          - url: "wss://192.168.1.101:36912"
```

In my testing, I've tried several different things. Sticky cookies, changing the URL from "wss" to "ws", or "http"/"https". Still cannot get it to connect for some reason. One reference I found on the web links to [here](https://community.traefik.io/t/websocker-to-different-port/10952/5), which I followed with no luck. 

I will test your latest changes and let you know.

**Comment by peanutlasko** [comment]
FYI - client side player works like a dream. There's a little lag, which is totally fine. 1 Problem solved!

I'm now working on the docker container/client side file browser. Will report back.

**Comment by Marekkon5** [comment]
For the file browser you should keep in mind, that 1T can only see the local filesystem - if you run it inside a container, it will see only the data in the container. You should mount your library inside your container to some accessible place (like `/library`), or just run 1T outside the container to have access to your entire library/files.

**Comment by peanutlasko** [comment]
Yep, just tested and now it's working! I've got my library mounted inside the container. 

Docker container is fully operational, thanks! Now just solving my reverse proxy issue and I should be all set. Much appreciated for this @Marekkon5 

I'll probably maintain the .dockerfile on my end which builds the container. If you want me to share, I'm happy to.

**Comment by Marekkon5** [comment]
The missing path error - it is caused by a default path fallback not being able to identify the default audio directories. This normally isn't a problem, because it will work in all normal desktops, however for docker containers you have to set manually the enviromnent variables: `XDG_MUSIC_DIR` to your default music dir and `HOME` to your home dir.

As for the reverse proxy - I have 0 experience with Trafeik and have no idea how to even make a basic setup with your config, and I won't be available next week, so I will take a look another day when I'm free.

**Comment by peanutlasko** [comment]
Thanks sir, I think I can fix the proxy issue on my own. I'll add those two env variables to my .dockerfile build and test again. Will keep you updated.

**Comment by flovanc** [comment]
hello @peanutlasko 

can you share your dockerfile, i expect to try 1t inside docker too ...

**Comment by peanutlasko** [comment]
Hey @flovanc , this is what I have so far. Still not able to get things working quite right though. I've set both the HOME and XDG_MUSIC_DIR but the container still throws an error about the missing path. 
```

FROM ubuntu:20.04

RUN apt update && apt install -y git
RUN DEBIAN_FRONTEND=noninteractive apt-get install -yq --no-install-recommends \
    autogen \
    pkg-config \
    make \
    gcc \
    g++ \
    libssl-dev \
    libsndfile1-dev \
    libwebkit2gtk-4.0-dev \
    libasound2-dev \
    curl && \ 
    apt-get clean

RUN mkdir /music

ENV XDG_MUSIC_DIR /music
ENV HOME /home/onetagger

WORKDIR /home/onetagger

RUN curl -L -o onetagger.tar.gz https://github.com/Marekkon5/onetagger/releases/download/1.4.0/OneTagger-linux.tar.gz && \
    tar zxvf onetagger.tar.gz

RUN ln -sf /root/.config/onetagger /config



CMD ./onetagger --server --expose
```

**Comment by flovanc** [comment]
Thanks @peanutlasko  

The error are maybe due to the fact that the dockerfile  download the 1.4.0 release and not the lattest build whit update from @Marekkon5  since two week ?

**Comment by peanutlasko** [comment]
Using a .dockerfile, I can only pull down from Github releases, not commits. I've tested using the latest [af417fb]commit by downloading the linux binary and replacing it in my container.

I get the new features like client side browser and client side playback. However, still getting the errors about missing paths

When I attach a shell and run "printenv HOME" or "printenv XDG_MUSIC_DIR" I get the correct path's displayed, so I know the variables are getting set. I suspect there's something else I need to do for the container regarding the directory paths that @Marekkon5 might know more about.

**Comment by Marekkon5** [comment]
The XDG_MUSIC_DIR part is the only one which can produce `Missing path!` from my code. However as I said - I won't be available this week, so I'll try to reproduce inside a container next week. Thanks

**Comment by Marekkon5** [comment]
Hello, in the latest commit I've fixed the `Missing path` error and the file browser in Edit Tags seems to be working fine. 

I've also added endpoint to the updater API to retrieve the Actions artifacts without logging into Github.
URL: `https://1t.marekkon5.workers.dev/artifact/COMMIT/ARTIFACT`
Replace `COMMIT` with the beginning of commit hash (for example `ead8380`) or use `latest` for the latest commit. Replace `ARTIFACT` with the artifact name (like `onetagger-linux` or `onetagger-win-setup`).

Example: `https://1t.marekkon5.workers.dev/artifact/latest/onetagger-linux`

Also I've modified your Docker image for my testings:

1. Some of the dependencies you've installed were only for building 1T, so they aren't neccessary for running it - I've removed them.
2. `unzip`, `ca-certificates`, and `curl` are only for downloading/installing you can `apt remove` them afterwards
3. I've created an actual user, so it can run without weird linking and permission stuff.
4. It uses the latest Actions artifact, rather than release.

```
FROM ubuntu:20.04

RUN apt update
RUN DEBIAN_FRONTEND=noninteractive apt install -yq --no-install-recommends \
    libssl-dev \
    git \
    libsndfile1-dev \
    libwebkit2gtk-4.0-dev \
    libasound2-dev \
    curl \
    ca-certificates \
    unzip
RUN apt clean

RUN mkdir /music

RUN useradd -ms /bin/bash onetagger
USER onetagger
WORKDIR /home/onetagger

RUN curl -L -o onetagger.zip https://1t.marekkon5.workers.dev/artifact/latest/onetagger-linux
RUN unzip onetagger.zip
RUN tar zxvf OneTagger-linux.tar.gz
RUN rm *.zip *.tar.gz

CMD ./onetagger --server --expose
```

Let me know if it works for you. Thanks

**Comment by flovanc** [comment]
hello and thanks for all the work on 1T ...

here is the error log afer the test on my ubuntu 18.04 dedicated server.

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }', crates/onetagger-shared/src/lib.rs:100:94
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by Marekkon5** [comment]
It seems like the config/log dir isn't writable. Are you running as correct user? Is the ~/.config/onetagger writable for that user? If you modified the Docker image, could you share it? Thanks

**Comment by flovanc** [comment]
ok add owner  and group and container start now ...

i try it and come back for some feedback.

**Comment by peanutlasko** [comment]
@Marekkon5 can confirm this latest fix and dockerfile works as expected. Have not run into any issues. I'll be switching to using the Docker container for my full time use. Still digging into my proxy issue but I'm making some progress. 

Thanks for all your help on this!

**Comment by flovanc** [comment]
> @Marekkon5 can confirm this latest fix and dockerfile works as expected. Have not run into any issues. I'll be switching to using the Docker container for my full time use. Still digging into my proxy issue but I'm making some progress.
> 
> Thanks for all your help on this!
> 
> Edit: Fixed the proxy issue as well. Looks like all working now!

hello @peanutlasko how have you fixed the proxy issue ? 1.5.0 work inside docker but can't use it whit traefik ...

**Comment by TheRealGramdalf** [comment]
@flovanc [My image](https://hub.docker.com/r/therealgramdalf/onetagger) over at dockerhub (using the tag `devel-ubuntu` should be working with a single port (`36913`).
